lib/index.ts

```diff
@@ -0,0 +1,572 @@
+type InternalCallback = (err: Error | null, response: any | null) => void;
+
+export class Front {
+	public comment: {
+		create: (params: FrontCommentRequest.Create, callback?: FrontCallback<FrontComment>) => Promise<FrontComment>;
+		get: (params: FrontCommentRequest.Get, callback?: FrontCallback<FrontComment>) => Promise<FrontComment>;
+		listMentions: (params: FrontCommentRequest.ListMentions, callback?: FrontCallback<FrontCommentMentions>)
+			=> Promise<FrontCommentMentions>;
+	};
+
+	public conversation: {
+		get: (params: FrontConversationRequest.Get, callback?: FrontCallback<FrontConversation>)
+			=> Promise<FrontConversation>;
+		list: (params?: FrontConversationRequest.List, callback?: FrontCallback<FrontConversations>)
+			=> Promise<FrontConversations>;
+		listComments: (params: FrontConversationRequest.ListComments, callback?: FrontCallback<FrontConversationComments>)
+			=> Promise<FrontConversationComments>;
+		listFollowers: (params: FrontConversationRequest.ListFollowers,
+			callback?: FrontCallback<FrontConversationFollowers>) => Promise<FrontConversationFollowers>;
+		listInboxes: (params: FrontConversationRequest.ListInboxes,
+			callback?: FrontCallback<FrontConversationInboxes>) => Promise<FrontConversationInboxes>;
+		listMessages: (params: FrontConversationRequest.ListMessages,
+			callback?: FrontCallback<FrontConversationMessages>) => Promise<FrontConversationMessages>;
+		update: (params: FrontConversationRequest.Update, callback?: FrontCallback<void>) => Promise<void>;
+	};
+
+	public inbox: {
+		create: (params: FrontInboxRequest.Create, callback?: FrontCallback<FrontInboxCreation>)
+			=> Promise<FrontInboxCreation>;
+		get: (params: FrontInboxRequest.Get, callback?: FrontCallback<FrontInbox>) => Promise<FrontInbox>;
+		list: (callback?: FrontCallback<FrontInboxes>) => Promise<FrontInboxes>;
+		listChannels: (params: FrontInboxRequest.ListChannels, callback?: FrontCallback<FrontInboxChannels>)
+			=> Promise<FrontInboxChannels>;
+		listConversations: (params: FrontInboxRequest.ListConversations,
+			callback?: FrontCallback<FrontInboxConversations>) => Promise<FrontInboxConversations>;
+		listTeammates: (params: FrontInboxRequest.ListTeammates, callback?: FrontCallback<FrontInboxTeammates>)
+			=> Promise<FrontInboxTeammates>;
+	};
+
+	public message: {
+		get: (params: FrontMessageRequest.Get, callback?: FrontCallback<FrontMessage>) => Promise<FrontMessage>;
+		send: (params: FrontMessageRequest.Send, callback?: FrontCallback<FrontConversationReference>)
+			=> Promise<FrontConversationReference>;
+		reply: (params: FrontMessageRequest.Reply, callback?: FrontCallback<FrontStatus>) => Promise<FrontStatus>;
+		receiveCustom: (params: FrontMessageRequest.ReceiveCustom,
+			callback?: FrontCallback<FrontConversationReference>) => Promise<FrontConversationReference>;
+	};
+
+	private apiKey: string;
+
+	constructor(apiKey: string) {
+		// Key.
+		this.apiKey = apiKey;
+
+		// Setup resource paths.
+		this.comment = {
```

Comment:
Well the first thing is that `httpCall` can be declared on the prototype level, the same as `formatPath` is, because whilst it does use `this.apiKey` that gets resolved at runtime at which point it will have it's `this` context be the instance of `Front` that it was called with, eg for `a.httpCall()` then `this` will be `a`, and for `b.httpCall()` then `this` will be `b` - you could also do `a.httpCall.call(b)` and the context would be `b` if you want to get really crazy.

The way I found for that was just to use
```typescript
	private httpCall(details: Request, params: any, callback?: InternalCallback): Promise<any | void> {
		const requestOpts = {
			body: (params) ? params : {},
			headers: {
				Authorization: `Bearer ${this.apiKey}`
			},
			json: true,
			method: details.reqType,
			url: `${URL}/${this.formatPath(details.path, params)}`
		};

		// Make the request.
		return request(requestOpts).promise().catch((error: any) => {
			// Format this into something useful, if we can.
			const frontError: FrontError = error.error._error;
			let newError: FrontError = Object.create(Error.prototype);
			if (frontError) {
				_.each([ 'status', 'title', 'message', 'details' ], (key) => {
					if (frontError[key]) {
						newError[key] = frontError[key]
					}
				});
			}

			throw newError;
		}).asCallback(callback);
	}
```
ie removing the `=`

As for getting the partials here to reference the prototyped `httpCall` within the typescript class syntax I'm not sure how, but the js equivalent would be either:
```js
create: function() {
     this.httpCall.apply(this, [{ reqType: 'POST', path: 'conversations/{conversation_id}/comments' }].concat(arguments))
}
```
or
```js
create: _.partial(Front.prototype.httpCall, { reqType: 'POST', path: 'conversations/{conversation_id}/comments' })
```